indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java

```diff
@@ -433,6 +438,76 @@ public static Path makeSegmentOutputPath(
     return outputPath;
   }
 
+  /**
+   * Rename the files. This works around some limitations of both FileContext (no s3n support) and NativeS3FileSystem.rename
+   * which will not overwrite
+   * @param outputFS The output fs
+   * @param indexZipFilePath The original file path
+   * @param finalIndexZipFilePath The to rename the original file to
+   * @return False if a rename failed, true otherwise (rename success or no rename needed)
+   * @throws IOException Bubbles up IOException from the hadoop classes.
+   */
+  private static boolean renameIndexFiles(
+      final FileSystem outputFS,
+      final Path indexZipFilePath,
+      final Path finalIndexZipFilePath
+  )
+      throws IOException
+  {
+    return ((DataPusher) RetryProxy.create(
+        DataPusher.class,
+        new DataPusher()
+        {
+          @Override
+          public long push() throws IOException
```

Comment:
it seems a little weird to use the DataPusher interface here. RetryUtils might be cleaner so we don't have to abuse the method to return a boolean.